### DISCUSSION TITLE
P2P without network IP

### DISCUSSION THREAD

**Discussion by fafarun** [kairos-io/kairos#1935] [Q&A] [answered]
HI,

I tried the mode p2p on RPI4, but I have some problem, maybe my use case is not good or my cloud-init file is wrong. 

*Use Case* : 
I generate a generic image for rpi4 with a cloud-init file. 
I connect my 3 rpi4 on a swith without another compute or service on this switch. 
I want my rpi4 start and create the edgevpn network and deploy k3s. 

*Implementation* : 
 my cloud-init : 

```
 #cloud-config
debug: true

hostname: metal-{{ trunc 4 .MachineID }}

users:
  - name: "kairos"                       
    passwd: "kairos"                      
    ssh_authorized_keys:                  
      - 'remove_sshd_key'
ssh_pwauth: False  
p2p:
  disable_dht: true 
  vpn:
    create: true 
    use: true 
    auto: true
    dns: true
    env:
      DHCP: "true"
      EDGEVPNDHT: "false"
      ADDRESS: "10.5.0.30/24"
  auto:
    enable: true
    ha:
      enable: true
      master_nodes: 2
  network_token: "remove_token"
```

*Result*  :

EdgeVPN start but it never try to see the other node on the network. 
This use case is the good one ? Or the node need an external ip ? 

I tried  2 others configuration : 

* the Switch is connected to my internet box. 

My 3 nodes got an ip with the dhcp on my internet box (192.168.0.x). After that the edgevpn gave a ip on each node (10.5.0.3x). 
K3s start. 

* After a first start without the switch and only the 3 nodes. 
After the first boot, edgevpn didn't try to reach The other node. 
I connect the switch to my internet box. I restarted the 3 nodes. 
the nodes got an ip on my Internet box (192.168.0.x), but the edgevpn did not work.
in the log the edge didn't had this ip in the node addresses.

**Comment by jimmykarily** [comment, marked as answer]
As far as I know, the nodes will try to discover each other using [mDNS](https://en.wikipedia.org/wiki/Multicast_DNS). For mDNS to work broadcasting messages should work which means the nodes should already belong to the same network.

In your setup, they are just plugged onto the same switch but without any service distributing IP addresses (DHCP) so they are not yet part of the same network. Broadcasting messages won't work, thus mDNS doesn't work.

When you plug your internet box, like you said, it distributes IP addresses thus everything works.

**Reply by fafarun** [reply]
Thanks for your answer. I thought it would be based on broadcast with macaddress. I need to read more about mDNS protocol.
